src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -1145,11 +1145,12 @@ abstract class RefChecks extends Transform {
     }
 
     // scala/bug#6276 warn for trivial recursion, such as `def foo = foo` or `val bar: X = bar`, which come up more frequently than you might think.
+    // scala/bug#11614 warns for trivial recursion around `def x(): X = x`, covering a method with an apply
     // TODO: Move to abide rule. Also, this does not check that the def is final or not overridden, for example
     def checkInfiniteLoop(sym: Symbol, rhs: Tree): Unit =
-      if (!sym.isValueParameter && sym.paramss.isEmpty) {
+      if (!sym.isValueParameter && (sym.paramss.isEmpty || sym.paramss.forall(_.isEmpty))) {
         rhs match {
-          case t@(Ident(_) | Select(This(_), _)) if t hasSymbolWhich (_.accessedOrSelf == sym) =>
+          case t@(Ident(_) | Select(This(_), _) | Apply(Select(This(_), _), _)) if t hasSymbolWhich (_.accessedOrSelf == sym) =>
```

Comment:
`t` is redundant because `rhs` is just a param. Note the outside parens were also redundant (as I learned recently).